glib/src/log.rs

```diff
@@ -328,27 +336,131 @@ pub fn log_default_handler(log_domain: Option<&str>, log_level: LogLevel, messag
+        )
+    }
+    // rustdoc-stripper-ignore-next
+    /// Creates a field from a raw pointer. Can be used to pass custom data into a writer function
+    /// set with [`log_set_writer_func`], where it can be retreived using [`value_ptr`]. Panics if
+    /// `key` contains null bytes.
+    pub fn new_ptr<T>(key: &'a GStr, ptr: *const T) -> Self {
+        Self(
+            ffi::GLogField {
+                key: key.as_ptr(),
+                value: ptr as *const _,
+                length: 0,
+            },
+            Default::default(),
+        )
+    }
+    // rustdoc-stripper-ignore-next
+    /// Retreives the field key.
+    pub fn key(&self) -> &str {
+        unsafe { std::ffi::CStr::from_ptr(self.0.key as *const i8) }
+            .to_str()
+            .unwrap()
+    }
+    // rustdoc-stripper-ignore-next
+    /// Retrieves the raw pointer of the field value. This should only be used on fields created
+    /// with [`new_ptr`] inside writer functions set with [`log_set_writer_func`].
+    pub fn value_ptr<T>(&self) -> *const T {
+        self.0.value as *const T
+    }
+    // rustdoc-stripper-ignore-next
+    /// Retrieves a byte array of the field value. Behavior is undefined if called on a field
+    /// created with [`new_ptr`].
+    pub fn value_bytes(&self) -> &[u8] {
+        if self.0.length < 0 {
+            unsafe { std::ffi::CStr::from_ptr(self.0.value as *const i8) }.to_bytes()
+        } else {
+            unsafe { std::slice::from_raw_parts(self.0.value as *const u8, self.0.length as usize) }
+        }
+    }
+    // rustdoc-stripper-ignore-next
+    /// Retrieves a string of the field value, or `None` if the string is not valid UTF-8. Behavior
+    /// is undefined if called on a field created with [`new_ptr`].
+    pub fn value_str(&self) -> Option<&str> {
+        std::str::from_utf8(self.value_bytes()).ok()
+    }
+}
+
+#[cfg(any(feature = "v2_50", feature = "dox"))]
+type WriterCallback = dyn Fn(LogLevel, &[LogField<'_>]) -> LogWriterOutput + Send + Sync + 'static;
+
+#[cfg(any(feature = "v2_50", feature = "dox"))]
+static WRITER_FUNC: once_cell::sync::OnceCell<Box<WriterCallback>> =
+    once_cell::sync::OnceCell::new();
+
+#[cfg(any(feature = "v2_50", feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(feature = "v2_50")))]
+#[doc(alias = "g_log_set_writer_func")]
+pub fn log_set_writer_func<
+    P: Fn(LogLevel, &[LogField<'_>]) -> LogWriterOutput + Send + Sync + 'static,
+>(
+    writer_func: P,
+) {
+    if WRITER_FUNC.set(Box::new(writer_func)).is_err() {
+        panic!("Writer func can only be set once");
+    }
+    unsafe extern "C" fn writer_trampoline(
+        log_level: ffi::GLogLevelFlags,
+        fields: *const ffi::GLogField,
+        n_fields: libc::size_t,
+        _user_data: ffi::gpointer,
+    ) -> ffi::GLogWriterOutput {
+        let writer_func = WRITER_FUNC.get().unwrap();
+        let fields = std::slice::from_raw_parts(fields as *const LogField<'_>, n_fields);
```

Comment:
If a field has an arbitrary pointer then usage of it from inside the log function could lead to UB within safe code. That doesn't seem ideal